-- local RagdollM = script.Parent
-- local ragdollCloneRE = RagdollM.RagdollCloneRE
-- local ragdollRE = RagdollM.RagdollRE
--
-- local camera = workspace.CurrentCamera :: Camera
--
-- ragdollCloneRE.OnClientEvent:Connect(function(humanoid: Humanoid)
--     camera.CameraSubject = humanoid
--     humanoid:SetStateEnabled(Enum.HumanoidStateType.GettingUp, false)
--     humanoid:SetStateEnabled(Enum.HumanoidStateType.FallingDown, false)
--     humanoid:ChangeState(Enum.HumanoidStateType.PlatformStanding)
-- end)
--
-- ragdollRE.OnClientEvent:Connect(function(humanoid: Humanoid, enabled: boolean)
--     humanoid:SetStateEnabled(Enum.HumanoidStateType.Jumping, enabled)
--     humanoid:SetStateEnabled(Enum.HumanoidStateType.GettingUp, enabled)
--     humanoid:SetStateEnabled(Enum.HumanoidStateType.FallingDown, enabled)
--     if not enabled then
--         humanoid:ChangeState(Enum.HumanoidStateType.PlatformStanding)
--     else
--         humanoid:ChangeState(Enum.HumanoidStateType.GettingUp)
--     end
-- end)
